### PR TITLE
Add cfunction tuple of types deprecation

### LINKED ISSUE #23020: Regression calling cfunction
```julia
using BenchmarkTools

foo(x) = x + 1
f() = cfunction(foo, Int, (Int,))

@btime f();
  815.538 ns (3 allocations: 96 bytes)

# version
# Julia Version 0.7.0-DEV.1159
# Commit 8d752a07f4* (2017-07-29 02:32 UTC)
# OS: Windows (x86_64-w64-mingw32)
```

Julia 0.6
```julia
using BenchmarkTools

foo(x) = x + 1
f() = cfunction(foo, Int, (Int,))

julia> @btime f();
  1.578 ns (0 allocations: 0 bytes)
```

### PR DESCRIPTION
**Pull request by musm**
close https://github.com/JuliaLang/julia/issues/23020

### CODE REVIEW

File: base/deprecated.jl [hunk 1]
```diff
@@ -1601,6 +1601,11 @@ end
 # issue #6466
 # `write` on non-isbits arrays is deprecated in io.jl.
 
+function cfunction(f, r, a::Tuple)
```
Comment (fredrikekre) [inline base/deprecated.jl]: could be simplified to
```julia
@deprecate cfunction(f, r, a::Tuple) cfunction(f, r, Tuple{a...})
```

Comment (musm) [inline base/deprecated.jl]: True, however I think the resulting deprecation warning message is not as helpful/clear in this case.

Comment (fredrikekre) [inline base/deprecated.jl]:
```julia
WARNING: cfunction(f, r, a::Tuple) is deprecated, use cfunction(f, r, Tuple{a...}) instead.
```
looks as helpful to me

**Comment by fredrikekre** [comment]
Did you lose the NEWS commit? I removed the label cause I saw you added that but now its not here?

**Comment by musm** [comment]
hmm that is super strange...I will add it back.

**Comment by musm** [comment]
Added back.

**Comment by tkelman** [comment]
ah wondered what happened there

**Comment by musm** [comment]
Github's rebase tools is strange. I had to rebase this twice, with the exact same changes to make it work.....

File: base/deprecated.jl [hunk 2]
```diff
@@ -1595,6 +1595,12 @@ end
 # issue #6466
 # `write` on non-isbits arrays is deprecated in io.jl.
 
+# PR #23066
```
Comment (fredrikekre) [inline base/deprecated.jl]: trailing white space

File: base/deprecated.jl (hunk 1 above)
Comment (tkelman) [inline base/deprecated.jl]: agreed, let the deprecate macro do what it's there for

**Comment by musm** [comment]
Unrelated Example install failures on appveyor  and strange (unrelated) timeouts  on travis

**Comment by fredrikekre** [comment]
Failure is definitely related:
```
Expression: stderr == "MethodError: no method matching this_function_has_no_methods()\n"
   Evaluated: "MethodError: no method matching this_function_has_no_methods()\nWARNING: cfunction(f, r, a::Tuple) is deprecated, use cfunction(f, r, Tuple{a...}) instead.\nStacktrace:\n [1] depwarn at ./deprecated.jl:70 [inlined]\n [2] cfunction(::Function, ::Type{T} where T, ::Tuple{}) at ./deprecated.jl:57\nwhile loading no file, in expression starting on line 0\n" == "MethodError: no method matching this_function_has_no_methods()\n"
Stacktrace:
 [1] macro expansion at /private/tmp/julia/share/doc/julia/examples/embedding/embedding-test.jl:20 [inlined]
 [2] macro expansion at ./test.jl:943 [inlined]
 [3] anonymous at ./<missing>:?
```

**Comment by iblis17** [comment]
:-o I should add the embedding testing to freebsd ci.

**Comment by musm** [comment]
@fredrikekre  good catch fixed!
@iblis17  👍

**Review by tkelman — APPROVED** [review]
unrelated download failure on linux32, timeout (all tests were done though) on linux64, so lgtm - should definitely be squashed

**Comment by musm** [comment]
g2g? (squash)

**Comment by musm** [comment]
merge ?

**Comment by musm** [comment]
squashed and rebased

**Comment by musm** [comment]
is this good to go now?

**Comment by KristofferC** [comment]
The docs need updating? https://docs.julialang.org/en/latest/manual/calling-c-and-fortran-code/#Creating-C-Compatible-Julia-Function-Pointers-1

**Comment by musm** [comment]
docs updated

**Comment by musm** [comment]
thanks @KristofferC

**Pull request merged**